napari/types.py

```diff
@@ -33,6 +42,36 @@
     Tuple[None, None, None],
 ]
 
+# these types are mostly "intentionality" placeholders.  While it's still hard
+# to use actual types to define what is acceptable data for a given layer,
+# these types let us point to a concrete namespace to indicate "this data is
+# intended to be (and is capable of) being turned into X layer type".
+# while their names should not change (without deprecation), their typing
+# implementations may... or may be rolled over to napari/image-types
+
+if tuple(np.__version__.split('.')) < ('1', '20'):
+    # this hack is because NewType doesn't allow `Any` as a base type
+    # and numpy <=1.20 didn't provide type stubs for np.ndarray
+    # https://github.com/python/mypy/issues/6701#issuecomment-609638202
+    class ArrayBase(np.ndarray):
+        def __getattr__(self, name: str) -> Any:
+            return super().__getattr__(name)
+
+
+else:
+    ArrayBase = np.ndarray  # type: ignore
+
+
+ImageData = NewType("ImageData", ArrayBase)
+LabelsData = NewType("LabelsData", ArrayBase)
+PointsData = NewType("PointsData", ArrayBase)
+ShapesData = NewType("ShapesData", List[ArrayBase])
+SurfaceData = NewType("SurfaceData", Tuple[ArrayBase, ArrayBase, ArrayBase])
+TracksData = NewType("TracksData", ArrayBase)
+VectorsData = NewType("VectorsData", ArrayBase)
+
+LayerDataTuple = NewType("LayerDataTuple", tuple)
```

Comment:
Is it possible to say something like `LayerDataTuple` is a tuple where the first element must be a `LayerData`, which must be one of the things above? Not needed, but curious how to think about that

Reply:
nope, not really... The rule is that the second argument to `NewType` _must_ be subclass-able.  So things like generic types (`Tuple[int, float...]`) don't work here.  These aren't type hints or type aliases, they are literally considered "new types" that derive from the their base types (except that there is no runtime checking and no overhead).  The "function" that is returned by `NewType` simply returns back whatever is passed to it: `lambda x: x`.  That's what I mostly mean by that comment "the implementation of these types may change over time, but their names should not).  For now, think of them as a concrete namespace for an "intention" that also happens to satisfy linters (i.e. if it says tuple, it at least has to be a tuple... but since there's no such `type` as "a tuple where the first element..." then we can't make it a `NewType`)

Reply:
yeah, ok that makes sense, matches what you said earlier, but thanks for the extra explanation!